// molli/packages/kind/index.js
const { Kind } = require('graphql/language')

module.exports = Object.assign(
  {},
  Kind,
  {
    TYPE_SCALAR: 'TypeScalar',
    TYPE_ENUM: 'TypeEnum',
    TYPE_UNION: 'TypeUnion',
    TYPE_INTERFACE: 'TypeInterface',
    TYPE_ARGUMENT: 'TypeArgument',
    TYPE_FIELD: 'TypeField',
    TYPE_OBJECT: 'TypeObject',
    TYPE_INPUT: 'TypeInput',
    TYPE_EXTENSION: 'TypeExtension',
    TYPE_PATH: 'TypePath',
    CRUD_CREATE_INPUT: 'CrudCreateInput',
    CRUD_CREATE_PAYLOAD: 'CrudCreatePayload',
    CRUD_UPDATE_INPUT: 'CrudUpdateInput',
    CRUD_UPDATE_PAYLOAD: 'CrudUpdatePayload',
    CRUD_DELETE_INPUT: 'CrudDeleteInput',
    CRUD_DELETE_PAYLOAD: 'CrudDeletePayload',
    CRUD_ORDER: 'CrudOrder',
    CRUD_FILTER: 'CrudFilter',
    CRUD_CONNECTION: 'CrudConnection',
    CREATION: 'Creation',
    UPDATE: 'Update',
    DELETE: 'Delete',
    FIND_ONE: 'FindOne',
    FIND_MULTIPLE: 'FindMultiple',
    RELATED_API: 'RelatedAPI',
    INT: 'Int',
    FLOAT: 'Float',
    STRING: 'String',
    ID: 'ID',
    BOOLEAN: 'Boolean',
    DATE: 'Date',
    ENUM: 'Enum',
    ORDER: 'Order',
  }
)
