// molli/packages/type/obj.js
const SimpleSchema = require('simpl-schema').default
const Field = require('./field')
const Path = require('./path')
const Kind = require('../kind')

class Obj {
  constructor(config) {
    const ConfigShape = new SimpleSchema({
      name: String,
      fields: {
        type: Array,
        optional: true,
        defaultValue: []
      },
      'fields.$': {
        type: Field,
        blackbox: true,
      },
      model: {
        type: Boolean,
        optional: true,
        defaultValue: false,
      },
      interfaces: {
        type: Array,
        optional: true,
        defaultValue: []
      },
      'interfaces.$': String,
    })
    ConfigShape.validate(config)
    Object.assign(
      this,
      { kind: Kind.TYPE_OBJECT },
      ConfigShape.clean(config)
    )
  }
  addField(field) {
    if (!field instanceof Field)
      throw new Error('field must be an instance of Field')
    this.fields.push(field)
  }
  addPaths(paths) {
    paths.forEach(path => {
      if (!path instanceof Path)
        throw new Error('path must be an instance of Path')
    })
    this.paths = paths
  }
}

module.exports = Obj
