// molli/packages/type/path.js
const SimpleSchema = require('simpl-schema').default
const Kind = require('../kind')
const Field = require('./field')

class Path {
  constructor(config) {
    const ConfigShape = new SimpleSchema({
      path: {
        type: Array,
        optional: true,
        defaultValue: [],
      },
      'path.$': {
        type: Field,
        blackbox: true,
      },
      field: {
        type: Field,
        blackbox: true,
      },
      defs: {
        type: Array,
        optional: true,
        defaultValue: [],
      },
      'defs.$': {
        type: Object,
        blackbox: true,
      },
      pass: Object,
      'pass.model': {
        type: Boolean,
        optional: true,
        defaultValue: false,
      },
      'pass.array': {
        type: Boolean,
        optional: true,
        defaultValue: false,
      },
      'pass.null': {
        type: Boolean,
        optional: true,
        defaultValue: false,
      },
      is: Object,
      'is.scalar': {
        type: Boolean,
        optional: true,
        defaultValue: false,
      },
      'is.enum': {
        type: Boolean,
        optional: true,
        defaultValue: false,
      },
      'is.union': {
        type: Boolean,
        optional: true,
        defaultValue: false,
      },
      'is.loop': {
        type: Boolean,
        optional: true,
        defaultValue: false,
      }
    })
    ConfigShape.validate(config)
    Object.assign(
      this,
      { kind: Kind.TYPE_PATH },
      ConfigShape.clean(config)
    )
  }
}

module.exports = Path
